src/uproot/interpretation/library.py

```diff
@@ -359,12 +359,18 @@ def _awkward_offsets(awkward, form, array):
 
 def _awkward_json_to_array(awkward, form, array):
     if form["class"] == "NumpyArray":
+        form = awkward.forms.from_json(json.dumps(form))
+        dtype = awkward.types.numpytype.primitive_to_dtype(form.primitive)
         if isinstance(array, awkward.contents.EmptyArray):
-            form = awkward.forms.from_json(json.dumps(form))
-            dtype = awkward.types.numpytype.primitive_to_dtype(form.primitive)
-            return awkward.contents.NumpyArray(numpy.empty(0, dtype=dtype))
+            return awkward.contents.NumpyArray(
+                numpy.empty(0, dtype=dtype),
+                parameters=form.parameters,
+            )
         else:
-            return array
+            return awkward.contents.NumpyArray(
+                numpy.asarray(array.data, dtype=dtype),
+                parameters=form.parameters,
+            )
```

Comment:
This corrects an error in the old, non-Forth method. The old method ran Python objects through `ak.from_iter` (which makes all Python integers `int64` and floats `float64`), then fixes them to agree with the dtypes in the actual Form. Many node types were fixed, but not non-empty leaf nodes. Now those are fixed as well.